In [1]:
import pandas as pd

In [2]:
draft_df = pd.read_csv('./data/College_WS.csv')

In [3]:
draft_df.head()

Unnamed: 0  SEASON  PLAYER_ID FIRST_NAME  LAST_NAME       PLAYER_NAME  \
0           0    2000       2124      malik      allen       Malik Allen   
1           1    2000      12019     harold  arceneaux  Harold Arceneaux   
2           2    2000      12020     lamont     barnes     Lamont Barnes   
3           3    2000      12131      mario      bland       Mario Bland   
4           4    2000       2056     primoz     brezec     Primoz Brezec   

  POSITION  HEIGHT_WO_SHOES HEIGHT_WO_SHOES_FT_IN  HEIGHT_W_SHOES  ...  \
0     PF-C            80.25             6' 8.25''             NaN  ...   
1    SG-SF            76.50              6' 4.5''             NaN  ...   
2     PF-C            80.50              6' 8.5''             NaN  ...   
3       PF            77.50              6' 5.5''             NaN  ...   
4        C            84.75             7' 0.75''             NaN  ...   

  OFF_DRIB_FIFTEEN_BREAK_LEFT  OFF_DRIB_FIFTEEN_TOP_KEY  \
0                         NaN                       NaN   
1                         NaN                       NaN   
2                         NaN                       NaN   
3                         NaN                       NaN   
4                         NaN                       NaN   

   OFF_DRIB_FIFTEEN_BREAK_RIGHT OFF_DRIB_COLLEGE_BREAK_LEFT  \
0                           NaN                         NaN   
1                           NaN                         NaN   
2                           NaN                         NaN   
3                           NaN                         NaN   
4                           NaN                         NaN   

   OFF_DRIB_COLLEGE_TOP_KEY OFF_DRIB_COLLEGE_BREAK_RIGHT  ON_MOVE_FIFTEEN  \
0                       NaN                          NaN              NaN   
1                       NaN                          NaN              NaN   
2                       NaN                          NaN              NaN   
3                       NaN                          NaN              NaN   
4                       NaN                          NaN              NaN   

   ON_MOVE_COLLEGE                Link  College WS  
0              NaN       malik-allen-1         4.2  
1              NaN  harold-arceneaux-1         5.6  
2              NaN     lamont-barnes-1         5.2  
3              NaN       mario-bland-1         4.2  
4              NaN     primoz-brezec-1         NaN  

[5 rows x 50 columns]

In [4]:
draft_df.isnull().sum()

Unnamed: 0                         0
SEASON                             0
PLAYER_ID                          0
FIRST_NAME                         0
LAST_NAME                          1
PLAYER_NAME                        0
POSITION                           5
HEIGHT_WO_SHOES                   44
HEIGHT_WO_SHOES_FT_IN             44
HEIGHT_W_SHOES                   347
HEIGHT_W_SHOES_FT_IN             347
WEIGHT                            46
WINGSPAN                          43
WINGSPAN_FT_IN                    43
STANDING_REACH                    45
STANDING_REACH_FT_IN              45
BODY_FAT_PCT                     276
HAND_LENGTH                      780
HAND_WIDTH                       780
STANDING_VERTICAL_LEAP           213
MAX_VERTICAL_LEAP                213
LANE_AGILITY_TIME                222
MODIFIED_LANE_AGILITY_TIME      1062
THREE_QUARTER_SPRINT             222
BENCH_PRESS                      457
SPOT_FIFTEEN_CORNER_LEFT        1448
SPOT_FIFTEEN_BREAK_LEFT         1446
S

In [5]:
pd.set_option('display.max_columns', None)
draft_df.head()

Unnamed: 0  SEASON  PLAYER_ID FIRST_NAME  LAST_NAME       PLAYER_NAME  \
0           0    2000       2124      malik      allen       Malik Allen   
1           1    2000      12019     harold  arceneaux  Harold Arceneaux   
2           2    2000      12020     lamont     barnes     Lamont Barnes   
3           3    2000      12131      mario      bland       Mario Bland   
4           4    2000       2056     primoz     brezec     Primoz Brezec   

  POSITION  HEIGHT_WO_SHOES HEIGHT_WO_SHOES_FT_IN  HEIGHT_W_SHOES  \
0     PF-C            80.25             6' 8.25''             NaN   
1    SG-SF            76.50              6' 4.5''             NaN   
2     PF-C            80.50              6' 8.5''             NaN   
3       PF            77.50              6' 5.5''             NaN   
4        C            84.75             7' 0.75''             NaN   

  HEIGHT_W_SHOES_FT_IN  WEIGHT  WINGSPAN WINGSPAN_FT_IN  STANDING_REACH  \
0                  NaN   271.0      86.5       7' 2.5''           109.0   
1                  NaN   219.0      80.5       6' 8.5''           103.0   
2                  NaN   235.5      87.5       7' 3.5''           108.0   
3                  NaN   287.0      84.0         7' 0''           103.0   
4                  NaN   243.0      86.0         7' 2''           110.0   

  STANDING_REACH_FT_IN  BODY_FAT_PCT  HAND_LENGTH  HAND_WIDTH  \
0               9' 1''           NaN          NaN         NaN   
1               8' 7''           NaN          NaN         NaN   
2               9' 0''           NaN          NaN         NaN   
3               8' 7''           NaN          NaN         NaN   
4               9' 2''           NaN          NaN         NaN   

   STANDING_VERTICAL_LEAP  MAX_VERTICAL_LEAP  LANE_AGILITY_TIME  \
0                    25.5               29.0              11.83   
1                     NaN               29.0              13.80   
2                    28.0               29.5              12.30   
3                    27.0               31.0              13.04   
4                    26.0               29.5              11.53   

   MODIFIED_LANE_AGILITY_TIME  THREE_QUARTER_SPRINT  BENCH_PRESS  \
0                         NaN                  3.38         13.0   
1                         NaN                   NaN          0.0   
2                         NaN                  3.40         10.0   
3                         NaN                  3.47         15.0   
4                         NaN                  3.55          NaN   

  SPOT_FIFTEEN_CORNER_LEFT SPOT_FIFTEEN_BREAK_LEFT SPOT_FIFTEEN_TOP_KEY  \
0                      NaN                     NaN                  NaN   
1                      NaN                     NaN                  NaN   
2                      NaN                     NaN                  NaN   
3                      NaN                     NaN                  NaN   
4                      NaN                     NaN                  NaN   

  SPOT_FIFTEEN_BREAK_RIGHT SPOT_FIFTEEN_CORNER_RIGHT SPOT_COLLEGE_CORNER_LEFT  \
0                      NaN                       NaN                      NaN   
1                      NaN                       NaN                      NaN   
2                      NaN                       NaN                      NaN   
3                      NaN                       NaN                      NaN   
4                      NaN                       NaN                      NaN   

  SPOT_COLLEGE_BREAK_LEFT SPOT_COLLEGE_TOP_KEY SPOT_COLLEGE_BREAK_RIGHT  \
0                     NaN                  NaN                      NaN   
1                     NaN                  NaN                      NaN   
2                     NaN                  NaN                      NaN   
3                     NaN                  NaN                      NaN   
4                     NaN                  NaN                      NaN   

  SPOT_COLLEGE_CORNER_RIGHT SPOT_NBA_CORNER_LEFT SPOT_NBA_BREAK_LEFT  \
0               

In [ ]:
# Remove columns from Bench Press to ON_MOVE_COLLEGE as there are a lot of null values and this data was not collected
# in early years of the NBA combine.

# Dropped these columns as there others that represent these: (Winspan,height with shoes, Standing Reach & Lane Agility time)

In [6]:
columns = ['WINGSPAN_FT_IN', 'HEIGHT_WO_SHOES', 'HEIGHT_WO_SHOES_FT_IN', 
        'HEIGHT_W_SHOES_FT_IN', 'STANDING_REACH_FT_IN', 'MODIFIED_LANE_AGILITY_TIME', 'Unnamed: 0', 'THREE_QUARTER_SPRINT']

draft_df.drop(draft_df.iloc[:, 24:49], inplace=True, axis=1)

draft_df.drop(draft_df.columns[[draft_df.columns.get_loc(col) for col in columns if col in draft_df]], 
              axis=1, inplace=True)

In [8]:
draft_df.head()

SEASON  PLAYER_ID FIRST_NAME  LAST_NAME       PLAYER_NAME POSITION  \
0    2000       2124      malik      allen       Malik Allen     PF-C   
1    2000      12019     harold  arceneaux  Harold Arceneaux    SG-SF   
2    2000      12020     lamont     barnes     Lamont Barnes     PF-C   
3    2000      12131      mario      bland       Mario Bland       PF   
4    2000       2056     primoz     brezec     Primoz Brezec        C   

   HEIGHT_W_SHOES  WEIGHT  WINGSPAN  STANDING_REACH  BODY_FAT_PCT  \
0             NaN   271.0      86.5           109.0           NaN   
1             NaN   219.0      80.5           103.0           NaN   
2             NaN   235.5      87.5           108.0           NaN   
3             NaN   287.0      84.0           103.0           NaN   
4             NaN   243.0      86.0           110.0           NaN   

   HAND_LENGTH  HAND_WIDTH  STANDING_VERTICAL_LEAP  MAX_VERTICAL_LEAP  \
0          NaN         NaN                    25.5               29.0   
1          NaN         NaN                     NaN               29.0   
2          NaN         NaN                    28.0               29.5   
3          NaN         NaN                    27.0               31.0   
4          NaN         NaN                    26.0               29.5   

   LANE_AGILITY_TIME  College WS  
0              11.83         4.2  
1              13.80         5.6  
2              12.30         5.2  
3              13.04         4.2  
4              11.53         NaN

In [9]:
draft_df.isnull().sum()

SEASON                      0
PLAYER_ID                   0
FIRST_NAME                  0
LAST_NAME                   1
PLAYER_NAME                 0
POSITION                    5
HEIGHT_W_SHOES            347
WEIGHT                     46
WINGSPAN                   43
STANDING_REACH             45
BODY_FAT_PCT              276
HAND_LENGTH               780
HAND_WIDTH                780
STANDING_VERTICAL_LEAP    213
MAX_VERTICAL_LEAP         213
LANE_AGILITY_TIME         222
College WS                241
dtype: int64

In [10]:
# Percentage of missing values in each column of the nba_draft_data.csv
draft_df.isnull().mean().mul(100).round(2).sort_values(ascending=False)


draft_df.isnull().mean()[draft_df.isnull().mean() !=0].mul(100)\
.round(2).sort_values(ascending=False) 

HAND_LENGTH               50.26
HAND_WIDTH                50.26
HEIGHT_W_SHOES            22.36
BODY_FAT_PCT              17.78
College WS                15.53
LANE_AGILITY_TIME         14.30
STANDING_VERTICAL_LEAP    13.72
MAX_VERTICAL_LEAP         13.72
WEIGHT                     2.96
STANDING_REACH             2.90
WINGSPAN                   2.77
POSITION                   0.32
LAST_NAME                  0.06
dtype: float64

In [11]:
# Looking at the rest of the null values I plan on replacing them with the mean per that columna and per position

draft_df.groupby("POSITION")["HAND_LENGTH"].mean()

POSITION
C        9.194030
C-PF     8.895833
PF       8.996644
PF-C     9.235294
PF-SF    8.886364
PG       8.291353
PG-SG    8.302632
SF       8.788462
SF-PF    8.732143
SF-SG    8.712500
SG       8.537338
SG-PG    8.557692
SG-SF    8.636905
Name: HAND_LENGTH, dtype: float64

I decided to replace the below listed columns with the average per position in that column as this would
give the most similar data for this null's.  This will allow me to not skew the dataset as replacing each
per postion will be the most similar to each player.

In [12]:
draft_df["HAND_LENGTH"].fillna(draft_df.groupby("POSITION")["HAND_LENGTH"].transform("mean"), inplace = True)
draft_df["HAND_WIDTH"].fillna(draft_df.groupby("POSITION")["HAND_WIDTH"].transform("mean"), inplace = True)
draft_df["BODY_FAT_PCT"].fillna(draft_df.groupby("POSITION")["BODY_FAT_PCT"].transform("mean"), inplace = True)
draft_df["HEIGHT_W_SHOES"].fillna(draft_df.groupby("POSITION")["HEIGHT_W_SHOES"].transform("mean"), inplace = True)
draft_df["STANDING_VERTICAL_LEAP"].fillna(draft_df.groupby("POSITION")["STANDING_VERTICAL_LEAP"].transform("mean"), inplace = True)
draft_df["MAX_VERTICAL_LEAP"].fillna(draft_df.groupby("POSITION")["MAX_VERTICAL_LEAP"].transform("mean"), inplace = True)
draft_df["LANE_AGILITY_TIME"].fillna(draft_df.groupby("POSITION")["LANE_AGILITY_TIME"].transform("mean"), inplace = True)




In [13]:
draft_df.isnull().sum()

SEASON                      0
PLAYER_ID                   0
FIRST_NAME                  0
LAST_NAME                   1
PLAYER_NAME                 0
POSITION                    5
HEIGHT_W_SHOES              4
WEIGHT                     46
WINGSPAN                   43
STANDING_REACH             45
BODY_FAT_PCT                5
HAND_LENGTH                 5
HAND_WIDTH                  5
STANDING_VERTICAL_LEAP      4
MAX_VERTICAL_LEAP           4
LANE_AGILITY_TIME           4
College WS                241
dtype: int64

In [14]:
draft_df.head(10)

SEASON  PLAYER_ID FIRST_NAME  LAST_NAME       PLAYER_NAME POSITION  \
0    2000       2124      malik      allen       Malik Allen     PF-C   
1    2000      12019     harold  arceneaux  Harold Arceneaux    SG-SF   
2    2000      12020     lamont     barnes     Lamont Barnes     PF-C   
3    2000      12131      mario      bland       Mario Bland       PF   
4    2000       2056     primoz     brezec     Primoz Brezec        C   
5    2000       2081     ernest      brown      Ernest Brown        C   
6    2000       2049     speedy    claxton    Speedy Claxton       PG   
7    2000      12132       eric      coley        Eric Coley    SG-SF   
8    2000      12133         ed       cota           Ed Cota       PG   
9    2000      12134      schea     cotton      Schea Cotton       SF   

   HEIGHT_W_SHOES  WEIGHT  WINGSPAN  STANDING_REACH  BODY_FAT_PCT  \
0       82.118103   271.0     86.50           109.0      9.297541   
1       78.064583   219.0     80.50           103.0      6.684507   
2       82.118103   235.5     87.50           108.0      9.297541   
3       81.066972   287.0     84.00           103.0      8.297265   
4       83.580108   243.0     86.00           110.0      9.239619   
5       83.580108   222.5     85.25           109.5      9.239619   
6       74.130693   166.0     72.00            94.5      6.420144   
7       78.064583   205.0     83.00           102.0      6.684507   
8       74.130693   189.0     76.00            96.5      6.420144   
9       79.708333   219.5     79.00           100.0      6.971429   

   HAND_LENGTH  HAND_WIDTH  STANDING_VERTICAL_LEAP  MAX_VERTICAL_LEAP  \
0     9.235294    9.875000               25.500000          29.000000   
1     8.636905    9.363095               29.761905          29.000000   
2     9.235294    9.875000               28.000000          29.500000   
3     8.996644    9.674497               27.000000          31.000000   
4     9.194030    9.936567               26.000000          29.500000   
5     9.194030    9.936567               27.053191          31.058511   
6     8.291353    8.971805               36.000000          42.500000   
7     8.636905    9.363095               35.000000          38.000000   
8     8.291353    8.971805               26.500000          33.500000   
9     8.788462    9.439904               34.000000          38.000000   

   LANE_AGILITY_TIME  College WS  
0           11.83000         4.2  
1           13.80000         5.6  
2           12.30000         5.2  
3           13.04000         4.2  
4           11.53000         NaN  
5           12.10766         NaN  
6           10.48000         7.0  
7           11.40000         7.0  
8           10.98000         2.9  
9           11.55000         2.0

Looking at the null values in the [College WS] column, there are 241 null values.  After doing research into these null 
values, I saw that the majority are null as they do not have this stat on the sports-reference.com.  This is due to the
fact that the player is:
    
    a.) Foreign 
    b.) old
    
Becasue of this I plan on dropping these values as it would impossible to re create these values without more data on their 
peformacne.

In [15]:

draft_df = draft_df.dropna()

In [16]:
draft_df.isnull().sum()

SEASON                    0
PLAYER_ID                 0
FIRST_NAME                0
LAST_NAME                 0
PLAYER_NAME               0
POSITION                  0
HEIGHT_W_SHOES            0
WEIGHT                    0
WINGSPAN                  0
STANDING_REACH            0
BODY_FAT_PCT              0
HAND_LENGTH               0
HAND_WIDTH                0
STANDING_VERTICAL_LEAP    0
MAX_VERTICAL_LEAP         0
LANE_AGILITY_TIME         0
College WS                0
dtype: int64

In [19]:
draft_df.to_csv('./data/Final_dataset_Cleaned.csv')

In [20]:
draft_df.shape

(1278, 17)